airflow/contrib/hooks/imap_hook.py

```diff
@@ -30,72 +34,105 @@ class ImapHook(BaseHook):
     """
     This hook connects to a mail server by using the imap protocol.
 
+    .. note:: Please call this Hook as context manager via `with`
+        to automatically open and close the connection to the mail server.
+
     :param imap_conn_id: The connection id that contains the information used to authenticate the client.
     :type imap_conn_id: str
     """
 
     def __init__(self, imap_conn_id='imap_default'):
         super().__init__(imap_conn_id)
-        self.conn = self.get_connection(imap_conn_id)
-        self.mail_client = imaplib.IMAP4_SSL(self.conn.host)
+        self.imap_conn_id = imap_conn_id
+        self.mail_client = None
 
     def __enter__(self):
-        self.mail_client.login(self.conn.login, self.conn.password)
-        return self
+        return self.get_conn()
 
     def __exit__(self, exc_type, exc_val, exc_tb):
         self.mail_client.logout()
 
-    def has_mail_attachment(self, name, mail_folder='INBOX', check_regex=False):
+    def get_conn(self):
+        """
+        Login to the mail server.
+
+        .. note:: Please call this Hook as context manager via `with`
+            to automatically open and close the connection to the mail server.
+
+        :return: an authorized ImapHook object.
+        :rtype: ImapHook
+        """
+
+        if not self.mail_client:
+            conn = self.get_connection(self.imap_conn_id)
+            self.mail_client = imaplib.IMAP4_SSL(conn.host)
+            self.mail_client.login(conn.login, conn.password)
+
+        return self
+
+    def has_mail_attachment(self, name, check_regex=False, mail_folder='INBOX', mail_filter='All'):
```

Comment:
How much do you want to re-order the args? I'm fine with this, but it means we can't have it in 1.10.x

Reply:
Another idea: since we are supporting py3 now we could do 
```suggestion
    def has_mail_attachment(self, name, *, check_regex=False, mail_folder='INBOX', mail_filter='All'):
```

To _force_ everything after name to be called via kwargs instead.

Reply:
That's a good idea. So 1.10.5 won't support py2 anymore?

Reply:
1.10 will still support py2.

Reply:
Okay. So what's actually left on the roadmap for 2.0? Will it definitely be released at end of this year or can it be earlier?

I don't need it yet - have it already in our custom plugin. So for me 2.0 would work. If any of you guys like this feature to be available earlier, let me know.

Reply:
You are likely the heaviest user of this, so we'll just merge this and make it 2.0 only

Reply:
Okay.